### PR TITLE
add SetBasic/MapBasic class

### PR DESCRIPTION
**Pull request by irexyc**
Add these classes for later use like `basic_subs`, `basic_solve_poly` etc.

### CODE REVIEW

**Comment by Marlin-Na** [comment]
@irexyc Sorry for the delay. Could you confirm the following problem on your computer? When I run `devtools::test()` multiple times, it will have the following output:

```
> devtools::test()
Loading symengine

***
*** Warning! The following Teuchos::RCPNode objects were created but have
*** not been destroyed yet.  A memory checking tool may complain that these
*** objects are not destroyed correctly.
***
*** There can be many possible reasons that this might occur including:
***
***   a) The program called abort() or exit() before main() was finished.
***      All of the objects that would have been freed through destructors
***      are not freed but some compilers (e.g. GCC) will still call the
***      destructors on static objects (which is what causes this message
***      to be printed).
***
***   b) The program is using raw new/delete to manage some objects and
***      delete was not called correctly and the objects not deleted hold
***      other objects through reference-counted pointers.
***
***   c) This may be an indication that these objects may be involved in
***      a circular dependency of reference-counted managed objects.
***

  0: RCPNode (map_key_void_ptr=0x75ba010)
       Information = {T=SymEngine::Integer, ConcreteT=SymEngine::Integer, p=0x75ba010, has_ownership=1}
       RCPNode address = 0x764b390
       insertionNumber = 560

NOTE: To debug issues, open a debugger, and set a break point in the function where
the RCPNode object is first created to determine the context where the object first
gets created.  Each RCPNode object is given a unique insertionNumber to allow setting
breakpoints in the code.  For example, in GDB one can perform:

1) Open the debugger (GDB) and run the program again to get updated object addresses

2) Set a breakpoint in the RCPNode insertion routine when the desired RCPNode is first
inserted.  In GDB, to break when the RCPNode with insertionNumber==3 is added, do:

  (gdb) b 'Teuchos::RCPNodeTracer::addNewRCPNode( [TAB] ' [ENTER]
  (gdb) cond 1 insertionNumber==3 [ENTER]

3) Run the program in the debugger.  In GDB, do:

  (gdb) run [ENTER]

4) Examine the call stack when the program breaks in the function addNewRCPNode(...)
Creating a new generic function for ‘cbind’ in package ‘symengine’
Creating a new generic function for ‘rbind’ in package ‘symengine’
SymEngine Version: 0.3.0
 _____           _____         _         
|   __|_ _ _____|   __|___ ___|_|___ ___ 
|__   | | |     |   __|   | . | |   | -_|
|_____|_  |_|_|_|_____|_|_|_  |_|_|_|___|
      |___|               |___|          
Testing symengine
```

Removing `test-set.R` will solves that. I will take a deeper look at it later.

File: src/set.cpp [hunk 1]
```diff
@@ -12,7 +12,7 @@ SEXP sexp_setbasic(SEXP ext) {
     CSetBasic*    set = elt_setbasic(out);
     CVecBasic*    vec = elt_vecbasic(ext);
     size_t        len = vecbasic_size(vec);
-    SEXP          a   = PROTECT(sexp_setbasic());
+    SEXP          a   = PROTECT(sexp_basic());
```
Comment (irexyc) [inline src/set.cpp]: I made the wrong type. And when to extract pointer, I didn't check its type first.
Should work fine now.

Comment (Marlin-Na) [reply src/set.cpp]: @irexyc Thanks a lot for fixing it.

**Comment by Marlin-Na** [comment]
Thanks a lot for your work, I will merge it for now.

There are a few things:

1. Set is unordered. Using `[` or `[[` would be confusing for users. I would rather suggest that directly use or expose `setbasic_get` and `setbasic_subset`.
2. From the user perspective, there is no need for `Set`. `VecBasic` would just works fine and also there is no concept of Set in R itself. However, `SetBasic` is used in some C functions. It might (or might not) be useful implementing them in the R level (if we do not have SetBasic on R level, we can just convert the results to VecBasic and argument to SetBasic in C upon function calling). But anyway, I think `SetBasic` should not be exposed to user, though we may use it internally.

**Pull request merged**